### ISSUE TITLE
thread '<unnamed>' panicked at 'Tried to unwrap Status::Partial'

### ISSUE THREAD

**Issue by divarvel** [steveklabnik/simple-server#80]
I get this error regularly on an application hosted on clever cloud: it's behind haproxy with a tcp check, and it's also pinged by zabbix (with a tcp check as well).

```
thread '<unnamed>' panicked at 'Tried to unwrap Status::Partial', /home/bas/.cargo/registry/src/github.com-1ecc6299db9ec823/httparse-1.2.3/src/lib.rs:254:31
2017-10-17T13:43:41.361Z: thread 'main' panicked at 'Thread pool worker panicked', /home/bas/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped_threadpool-0.1.8/src/lib.rs:236:12
```

I haven't been able to reproduce this exact error locally (with only haproxy), but I got this one instead:

```
thread '<unnamed>' panicked at 'Error handling connection.: Io(Error { repr: Os { code: 54, message: "Connection reset by peer" } })', src/libcore/result.rs:860:4
```

**Comment by steveklabnik** [comment]
Could you give us a stacktrace? Set `RUST_BACKTRACE` to anything. Thanks!

**Comment by ashleygwilliams** [comment]
this *might* be related to https://github.com/steveklabnik/simple-server/pull/76? if you upgrade to 0.2.0, @divarvel, does it still panic this way? without the stack trace it's hard to tell, but seems possible? anyways, let us know!

**Comment by divarvel** [comment]
This test was made with `0.2.0`

The stacktrace on my local test is the following:

```
thread '<unnamed>' panicked at 'Error handling connection.: Io(Error { repr: Os { code: 54, message: "Connection reset by peer" } })', src/libcore/result.rs:860:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: simple_server::Server::listen::{{closure}}::{{closure}}
  11: <F as scoped_threadpool::FnBox>::call_box
  12: scoped_threadpool::Pool::new::{{closure}}
thread 'main' panicked at 'Thread pool worker panicked', /Users/clementd/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped_threadpool-0.1.8/src/lib.rs:236:12
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: scoped_threadpool::Scope::join_all
   6: <scoped_threadpool::Scope<'pool, 'scope> as core::ops::drop::Drop>::drop
   7: core::ptr::drop_in_place
   8: scoped_threadpool::Pool::scoped
   9: simple_server::Server::listen
  10: clementd_win::main
  11: __rust_maybe_catch_panic
  12: std::rt::lang_start
  13: main
```

On clever cloud (the `thread '<unnamed>' panicked at 'Tried to unwrap Status::Partial'` error), I can't get a stracktrace yet, as there is an issue with rust on exherbo (@keruspe is working on it though)

**Comment by ashleygwilliams** [comment]
oh interesting! thanks for the clarification @divarvel ! we'll take a look at this!

**Comment by divarvel** [comment]
if you want to reproduce it locally, here's the haproxy config I've used https://gist.github.com/divarvel/a2ce0c074550bff0aa2f73792bb67d33

**Comment by gmbeard** [comment]
I think I may know what's causing this...

`handle_connection` only attempts one read of the stream. It's possible that the read doesn't consume the entire request (e.g. over a slow connection), which would explain the `Partial` result. This is the code I'm referring to...

        if stream.read(&mut buffer)? == 0 {
            // Connection closed
            return Ok(());
        }

        let request = parse_request(&buffer)?;

This issue won't necessarily show up using the examples on `127.0.0.1`, but it can be simulated by using telnet to the `server` example and just attempting to type in `GET / HTTP/1.1` etc. by hand; It quickly panics with a similar error to @divarvel 's.

To resolve this, the stream should be read in a loop until `parse_request` returns a valid request. However, there's going to be a fight with the borrow checker if a loop is used because a `Request` just borrows the bytes of the buffer which the stream is read into; Once `parse_request` is called, `buffer` cannot be mutated.

**Comment by steveklabnik** [comment]
Ah yes, that makes perfect sense.

Given that hyper uses httparse, this must be possible somehow. Will have to investigate more!

**Comment by gmbeard** [comment]
Are you open to contributions? I have a potential fix for this in a fork. Should I submit a PR? (Sorry, I don’t GitHub very well)

**Comment by ashleygwilliams** [comment]
@gmbeard totally! we have some requirements about keeping the example interface the same, and some code readability stuff but please do open a PR and we can start a conversation there. probably worth giving our `CONTRIBUTING.md` a glance too :)

**Comment by divarvel** [comment]
I've tested @gmbeard's branch (https://github.com/steveklabnik/simple-server/pull/84) and it solves the issue indeed (both locally and on Clever Cloud).

**Comment by dwrensha** [comment]
I was hitting this problem with simple-server 0.3.0, running behind a Google Cloud HTTPS Load Balancer. After I updated to the latest git version (https://github.com/steveklabnik/simple-server/commit/83540213e137240077e7f17e81203cf47601b367) the panic went away, but the server still did not work. If I remember correctly, it just failed to respond to my requests. I ended up switching to rocket.rs, which did not exhibit this problem.

**Comment by steveklabnik** [comment]
I've merged #84, so I think I'm going to give this a close. 

@dwrensha sorry to hear that! Since you've switched, I guess it's not a huge deal, but if you have any way of reproducing and still want to follow through, please open a new bug. Thanks!

**Comment by nyarly** [comment]
I am still running into this - it came up at today's RustBridge exercises. I suspect that my network stack surfaces the partial streaming more reliably?

**Comment by nyarly** [comment]
I worked out that a) I was using the older (published) 0.3 version of this crate, and that building from mainline git works.

I'm on Linux 4.14, though, and simple-server exhibits the "Status::Partial" panic 100% of the time on first request for me. A "hello world" server is enough to exhibit the problem.
